### PR TITLE
Add fourier potential

### PR DESCRIPTION
**Pull request by zhou-pj**
## Description

Add Fourier potential

## Motivation and Context
Add a Fourier potential that is optimized with sincos function and variable reuse

## How Has This Been Tested?

Included a test code that compare force and energy for a two particle system against python calculated value

## Change log

```
Add Fourier Potential
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/hoomd-blue/blob/master/sphinx-doc/credits.rst).

### CODE REVIEW

**Comment by zhou-pj** [comment]
HI @joaander and @jglaser. I made an improved Fourier potential based on your recommendations. I appreciate any feedback on the draft and please let me know if you find any issue with it. Thanks a lot!

File: hoomd/md/EvaluatorPairFourier.h [hunk 1]
```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2009-2016 The Regents of the University of Michigan
```
Comment (joaander) [inline hoomd/md/EvaluatorPairFourier.h]: 2009-2019

File: hoomd/md/FourierDriverPotentialPairGPU.cu [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) 2009-2018 The Regents of the University of Michigan
```
Comment (joaander) [inline hoomd/md/FourierDriverPotentialPairGPU.cu]: 2009-2019

File: hoomd/md/EvaluatorPairFourier.h [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+// TODO: documentation
+//! Class for evaluating the Fourier pair potential
+/*! <b>General Overview</b>
+
+    See EvaluatorPairLJ
+
+    <b>Fourier specifics</b>
+
+    EvaluatorPairFourier evaluates the function:
+
```
Comment (joaander) [inline hoomd/md/EvaluatorPairFourier.h]: Please document the class more thoroughly.

File: hoomd/md/EvaluatorPairFourier.h [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        //! Define the parameter type used by this pair potential evaluator
+        typedef pair_fourier_params param_type; //first try a 4th order fourier expression of potential
+        //! Constructs the pair potential evaluator
+        /*! \param _rsq Squared distance beteen the particles
+            \param _rcutsq Sqauared distance at which the potential goes to 0
+            \param _params Per type pair parameters of this potential
+        */
+        DEVICE EvaluatorPairFourier(Scalar _rsq, Scalar _rcutsq,
+                                    const param_type& _params)
+            {
```
Comment (joaander) [inline hoomd/md/EvaluatorPairFourier.h]: The compiler might not optimize out the copy into `a[]` and `b[]`. I recommend declaring a class member variable `const pair_fourier_params& params`, initialized with `: params(_params)` here.

File: hoomd/md/EvaluatorPairFourier.h [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+                Scalar x = r * period_scale;
+                Scalar r1inv = Scalar(1)/r;
+                Scalar r2inv = Scalar(1)/rsq;
+                Scalar r3inv = r1inv * r2inv;
+                Scalar r12inv = r3inv * r3inv * r3inv * r3inv;
+                Scalar a1 = 0;
+                Scalar b1 = 0;
+                for (int i=2; i<5; i++)
+                    {
+                    a1 = a1 + std::pow(-1,i) * a[i-2];
```
Comment (joaander) [inline hoomd/md/EvaluatorPairFourier.h]: Use `fast::pow`

File: hoomd/md/EvaluatorPairFourier.h [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+            }
+
+        #ifndef NVCC
+        //! Get the name of this potential
+        /*! \returns The potential name. Must be short and all lowercase, as this is the name energies will be logged as
+            via analyze.log.
+        */
+        static std::string getName()
+            {
+            return std::string("Fourier");
```
Comment (joaander) [inline hoomd/md/EvaluatorPairFourier.h]: potentials names should be all lowercase

File: hoomd/md/pair.py [hunk 7]
```diff
@@ -2764,3 +2764,73 @@ def process_coeff(self, coeff):
         lj1 = 4.0 * epsilon * math.pow(sigma, 12.0);
         lj2 = alpha * 4.0 * epsilon * math.pow(sigma, 8.0);
         return _hoomd.make_scalar2(lj1, lj2);
+
+class Fourier(pair):
```
Comment (joaander) [inline hoomd/md/pair.py]: potential names should be all lowercase

File: hoomd/md/pair.py [hunk 8]
```diff
@@ -2764,3 +2764,73 @@ def process_coeff(self, coeff):
...
+
+    :py:class:`Fourier` specifies that a Oscillating pair potential with fast decay near cutoff should be applied between every
+    non-excluded particle pair in the simulation.
+
+    .. math::
+        :nowrap:
+
+        \begin{eqnarray*}
+
+        # TODO add formula
```
Comment (joaander) [inline hoomd/md/pair.py]: Please document the potential.

File: hoomd/md/pair.py [hunk 9]
```diff
@@ -2764,3 +2764,73 @@ def process_coeff(self, coeff):
...
+            self.nlist.cpp_nlist.setStorageMode(_md.NeighborList.storageMode.full);
+            self.cpp_force = _md.PotentialPairFourierGPU(hoomd.context.current.system_definition, self.nlist.cpp_nlist, self.name);
+            self.cpp_class = _md.PotentialPairFourierGPU;
+
+        hoomd.context.current.system.addCompute(self.cpp_force, self.force_name);
+
+        # setup the coefficent options
+
+        self.required_coeffs = ['fourier_a','fourier_b'];
+        # self.pair_coeff.set_default_coeff('alpha', 1.0);
```
Comment (joaander) [inline hoomd/md/pair.py]: remove commented code that is not used

**Pull request merged**